src/Aspire.Hosting/Dashboard/DashboardViewModelService.cs

```diff
@@ -294,6 +321,90 @@ private async Task WriteChange(ResourceViewModel resourceViewModel, ObjectChange
+                // the endpoint is ready
+                var content = await response.Content.ReadAsStringAsync(_cancellationToken).ConfigureAwait(false);
+                if (!string.Equals(content, errorContent, StringComparison.Ordinal))
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                // Endpoint returned a valid response
+                return true;
+            }
+        }
+        catch
+        {
+            // Catch exceptions, we'll retry again
+        }
+
+        return false;
+    }
+
+    // Poll a given service endpoint until it doesn't return a bad 404 response from the proxy (or the attempt times out)
+    private async Task PollUntilServiceEndpointValid(WatchEventType eventType, Service service, TimeSpan maxRetryDuration)
+    {
+        try
+        {
+            const string errorContent = "404 page not found\n";
+
+            if (service.UsesHttpProtocol(out var uriScheme)
+                && !_checkedServices.Contains(service.Metadata.Name))
+            {
+                // We need to verify if service is "actually" working
+                if (!string.Equals(service.Status?.State, "Ready", StringComparison.OrdinalIgnoreCase))
+                {
+                    // If service is not ready, we don't try to connect. We will get future notification that service moved to ready state
+                    return;
+                }
+
+                var currentTimestamp = DateTime.UtcNow;
+                var delay = TimeSpan.FromMilliseconds(40);
+                var proxyUrlString = $"{uriScheme}://{service.AllocatedAddress}:{service.AllocatedPort}";
+                while (true)
+                {
+                    if (await CheckServiceEndpointReady(proxyUrlString, errorContent).ConfigureAwait(false))
+                    {
+                        break;
+                    }
+
+                    if (DateTime.UtcNow.Subtract(currentTimestamp) > maxRetryDuration)
+                    {
+                        // We went over max retry duration so exit while
+                        _logger.LogInformation("Couldn't confirm {ServiceName} endpoint ready in {TimeoutSeconds}, assuming ready", service.Metadata.Name, maxRetryDuration.TotalSeconds);
+                        break;
+                    }
+
+                    await Task.Delay(delay, _cancellationToken).ConfigureAwait(false);
+                    delay *= 2;
+                }
+
+                _checkedServices.Add(service.Metadata.Name);
+            }
+
+            await _kubernetesChangesChannel.Writer.WriteAsync(
+                (eventType, service.Metadata.Name, service), _cancellationToken).ConfigureAwait(false);
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Failed to confirm {ServiceName} endpoint ready", service.Metadata.Name);
```

Comment:
I don't think we should log this as error, we should log it as debug. This might show up by default and people will file bugs.